gluten-core/src/main/scala/org/apache/gluten/extension/columnar/validator/Validators.scala

```diff
@@ -0,0 +1,212 @@
+      buffer += new FallbackComplexExpressions(conf.fallbackExpressionsThreshold)
+      this
+    }
+
+    /** Fails validation on non-scan plan nodes if Gluten is running as scan-only mode. */
+    def fallbackIfScanOnly(): Builder = {
+      buffer += new FallbackIfScanOnly(conf.enableScanOnly)
+      this
+    }
+
+    /**
+     * Fails validation if native-execution of a plan node is not supported by current backend
+     * implementation by checking the active BackendSettings.
+     */
+    def fallbackByBackendSettings(): Builder = {
+      buffer += new FallbackByBackendSettings(settings)
+      this
+    }
+
+    /**
+     * Fails validation if native-execution of a plan node is disabled by Gluten/Spark
+     * configuration.
+     */
+    def fallbackByUserOptions(): Builder = {
+      buffer += new FallbackByUserOptions(conf)
+      this
+    }
+
+    /** Add a custom validator to pipeline. */
+    def add(validator: Validator): Builder = {
+      buffer += validator
+      this
+    }
+
+    def build(): Validator = {
+      if (buffer.isEmpty) {
+        NoopValidator
+      } else {
+        new ValidatorPipeline(buffer)
+      }
+    }
+  }
+
+  private object Builder {
+    def apply(): Builder = new Builder()
+  }
+
+  private object FallbackByHint extends Validator {
+    override def validate(plan: SparkPlan): Validator.OutCome = {
+      if (TransformHints.isNotTransformable(plan)) {
+        val hint = TransformHints.getHint(plan).asInstanceOf[TRANSFORM_UNSUPPORTED]
+        return fail(hint.reason.getOrElse("Reason not recorded"))
+      }
+      pass()
+    }
+  }
+
+  private class FallbackComplexExpressions(threshold: Int) extends Validator {
+    override def validate(plan: SparkPlan): Validator.OutCome = {
+      if (plan.expressions.exists(e => ExpressionUtils.getExpressionTreeDepth(e) > threshold)) {
+        return fail(
+          s"Disabled because at least one present expression exceeded depth threshold: " +
+            s"${plan.nodeName}")
+      }
+      pass()
+    }
+  }
+
+  private class FallbackIfScanOnly(scanOnly: Boolean) extends Validator {
+    override def validate(plan: SparkPlan): Validator.OutCome = plan match {
+      case _: BatchScanExec => pass()
+      case _: FileSourceScanExec => pass()
+      case p if HiveTableScanExecTransformer.isHiveTableScan(p) => pass()
+      case p if scanOnly => fail(p)
+      case _ => pass()
+    }
+  }
+
+  private class FallbackByBackendSettings(settings: BackendSettingsApi) extends Validator {
+    override def validate(plan: SparkPlan): Validator.OutCome = plan match {
+      case p: ShuffleExchangeExec if !settings.supportColumnarShuffleExec() => fail(p)
+      case p: SortMergeJoinExec if !settings.supportSortMergeJoinExec() => fail(p)
+      case p: WriteFilesExec
+          if !(settings.enableNativeWriteFiles() && settings.supportTransformWriteFiles) =>
+        fail(p)
+      case p: SortAggregateExec if !settings.replaceSortAggWithHashAgg =>
```

Comment:
Looks no need to keep `replaceSortAggWithHashAgg` in backend api now. Maybe, move the check (depends on config) for `SortAggregateExec` to `FallbackByUserOptions`?

If it makes sense, it's ok to do the refactor in follow-up pr.

Reply:
I got your point except that it seems to be an individual patch indeed. This PR just moves the options/settings checks around without optimizing them. Let's do that in another patch.